### PR TITLE
Integration testing - cluster security TLS

### PR DESCRIPTION
**Pull request by tohar-ayash**
Signed-off-by: tohar-ayash <toharayash1@gmail.com>

### CODE REVIEW

File: test/cluster/basic_test.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ func TestBasicCluster(t *testing.T) {
 			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
 		},
 	}
-	c, err := setup.NewCluster(setupConfig)
+	c, err := setup.NewCluster(setupConfig, false)
```
Comment (tock-ibm) [inline test/cluster/basic_test.go:65]: Put the TLS config options (say, clusterTLSEnabled=false/true) into the the setup.Config object... it will be false by default, so all unrelated tests stay the same.

File: test/cluster/security_test.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/hyperledger-labs/orion-server/test/setup"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
```
Comment (tock-ibm) [inline test/cluster/security_test.go:19]: goimports

File: test/cluster/security_test.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+)
+
+// Scenario:
+// - start 3 servers in a cluster with TLS enabled.
+// - stop the leader.
+// - submit a tx to the two remaining servers.
```
Comment (tock-ibm) [inline test/cluster/security_test.go:25]: submit multiple txs

File: test/cluster/security_test.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+		return newLeader >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	txID, rcpt, err := c.Servers[newLeader].WriteDataTx(t, worldstate.DefaultDBName, "bob", []byte("bob-data"))
+	require.NoError(t, err)
+	require.NotNil(t, rcpt)
+	require.True(t, txID != "")
+	require.True(t, len(rcpt.GetHeader().GetValidationInfo()) > 0)
+	require.Equal(t, types.Flag_VALID, rcpt.Header.ValidationInfo[rcpt.TxIndex].Flag)
+	t.Logf("tx submitted: %s, %+v", txID, rcpt)
```
Comment (tock-ibm) [inline test/cluster/security_test.go:75]: write more than one, say 10

File: test/cluster/security_test.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+	clusterLogger.Info("Snap list: ", snapList)
+}
+
+// Scenario:
+// - start 3 servers in a cluster with TLS enabled.
+// - stop the leader.
+// - wait for new server to be the leader.
+// - the leader submit a tx.
+// - repeat until each server was a leader at least once.
+// - make sure each node submitted a valid tx as a leader.
```
Comment (tock-ibm) [inline test/cluster/security_test.go:252]: You can drop this scenario as it does not add much to the security testing.

Instead, add a scenario in which we test negative cases: 
- a server with TLS disabled cannot connect to the cluster
- a server with TLS enabled but a bad certificate either will not start or won't connect
- a server with TLS enabled but a key-cert pair not from the cluster's CA will not connect
etc.

That is, start a cluster with 3 servers, in which one is misconfigured as indicated above, start the cluster, make sure the correct 2 do not connect to the misconfigured one.

You can do all the negative tests in another PR, "TLS negative tests", this way we'll move faster.

File: test/setup/cluster_setup.go [hunk 6]
```diff
@@ -48,7 +48,7 @@ type Config struct {
 }
 
 // NewCluster creates a new cluster environment for the blockchain database
-func NewCluster(conf *Config) (*Cluster, error) {
+func NewCluster(conf *Config, enableTLS bool) (*Cluster, error) {
```
Comment (tock-ibm) [inline test/setup/cluster_setup.go:51]: put `enableTLS` in the `conf` object

File: test/cluster/security_test.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+		return leaderIndex >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	// create server with TLS disabled
+	configEnv, err := c.Servers[leaderIndex].QueryConfig(t, "admin")
+	require.NoError(t, err)
+	require.NotNil(t, configEnv)
+	newServer := addServerTx(t, c, setupConfig, configEnv, leaderIndex, 3, false)
+	require.NotNil(t, newServer)
+	require.EqualError(t, c.StartServer(newServer), "failed to start the server: node-4")
```
Comment (tock-ibm) [inline test/cluster/security_test.go:292]: The server should try to connect forever, not fail

File: test/cluster/security_test.go [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+	require.NoError(t, server5.CreateConfigFile(true))
+
+	configEnv, err = c.Servers[leaderIndex].QueryConfig(t, "admin")
+	require.NoError(t, err)
+	require.NotNil(t, configEnv)
+
+	node5.Certificate = []byte("invalidCert")
+	newConfig := configEnv.GetResponse().GetConfig()
+	newConfig.ConsensusConfig.Members = append(newConfig.ConsensusConfig.Members, peer5)
+	newConfig.Nodes = append(newConfig.Nodes, node5)
```
Comment (tock-ibm) [inline test/cluster/security_test.go:310]: In the local config, put a second CA from which you create a second certificate.
The server will be rejected by the cluster peers 1,2,3 and will try to communicate forever.

File: test/cluster/security_test.go [hunk 9]
```diff
@@ -0,0 +1,435 @@
+// Copyright IBM Corp. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package cluster
+
+import (
+	"fmt"
+	"github.com/hyperledger-labs/orion-server/config"
+	"github.com/hyperledger-labs/orion-server/pkg/server/testutils"
```
Comment (tock-ibm) [inline test/cluster/security_test.go:9]: goimports

File: test/cluster/security_test.go [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	//require.NoError(t, c.StartServer(c.Servers[0]))
+	//require.NoError(t, c.StartServer(c.Servers[1]))
```
Comment (tock-ibm) [inline test/cluster/security_test.go:386]: remove comment

File: test/cluster/security_test.go [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+	require.NoError(t, c.Start())
```
Comment (tock-ibm) [inline test/cluster/security_test.go:387]: you are starting all the cluster, so wait for 0,1,2 to get a common leader, then shutdown 2 and continue, i.e. shutdown 2, wait for 0,1 leader, changing the crypto on 2.

File: test/cluster/security_test.go [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+	require.True(t, txID != "")
+	require.True(t, len(rcpt.GetHeader().GetValidationInfo()) > 0)
+	require.Equal(t, types.Flag_VALID, rcpt.Header.ValidationInfo[rcpt.TxIndex].Flag)
+	t.Logf("tx submitted: %s, %+v", txID, rcpt)
+
+	// server2 is not sync with the last tx, because it is unable to connect to the cluster
+	dataEnv, err := c.Servers[2].QueryData(t, worldstate.DefaultDBName, "key1", "admin")
+	require.NoError(t, err)
+	require.NotNil(t, dataEnv)
+	require.Equal(t, []byte(nil), dataEnv.GetResponse().GetValue())
```
Comment (tock-ibm) [inline test/cluster/security_test.go:433]: we can also do an eventually that server 2 has no leader, like what you fixed on node removal, and that the leader has only 2 active

File: test/cluster/security_test.go [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+	require.NoError(t, c.StartServer(s2))
+
+	// submit tx
+	txID, rcpt, _, err := c.Servers[leaderIndex].WriteDataTx(t, worldstate.DefaultDBName, "key1", []byte("key1-data"))
+	require.NoError(t, err)
+	require.NotNil(t, rcpt)
+	require.True(t, txID != "")
+	require.True(t, len(rcpt.GetHeader().GetValidationInfo()) > 0)
+	require.Equal(t, types.Flag_VALID, rcpt.Header.ValidationInfo[rcpt.TxIndex].Flag)
+	t.Logf("tx submitted: %s, %+v", txID, rcpt)
```
Comment (tock-ibm) [inline test/cluster/security_test.go:427]: also do eventually on the height of 0,1

**Review by tock-ibm — COMMENTED** [review]
some minor comments this time, looks good overall

**Review by tock-ibm — APPROVED** [review]
